flink-runtime/src/main/java/org/apache/flink/runtime/state/SnapshotDirectory.java

```diff
@@ -0,0 +1,172 @@
+	}
+
+	public boolean mkdirs() throws IOException {
+		return fileSystem.mkdirs(directory);
+	}
+
+	@Nonnull
+	public FileSystem getFileSystem() {
+		return fileSystem;
+	}
+
+	public boolean exists() throws IOException {
+		return fileSystem.exists(directory);
+	}
+
+	/**
+	 * List the statuses of the files/directories in the snapshot directory.
+	 *
+	 * @return the statuses of the files/directories in the given path.
+	 * @throws IOException if there is a problem creating the file statuses.
+	 */
+	public FileStatus[] listStatus() throws IOException {
+		return fileSystem.listStatus(directory);
+	}
+
+	/**
+	 * Calling this method completes the snapshot into the snapshot directory and creates a corresponding
+	 * {@link DirectoryStateHandle} that points to the snapshot directory. Calling this method will also change the
+	 * lifecycle state from "ongoing" to "completed". If the state was already deleted, an {@link IOException} is
+	 * thrown.
+	 *
+	 * @return a directory state handle that points to the snapshot directory.
+	 * @throws IOException if the state of this snapshot directory object is different from "ongoing".
+	 */
+	public DirectoryStateHandle completeSnapshotAndGetHandle() throws IOException {
+		if (state.compareAndSet(State.ONGOING, State.COMPLETED)) {
+			return new DirectoryStateHandle(directory, fileSystem);
+		} else {
+			throw new IOException("Expected state " + State.ONGOING + " but found state " + state.get());
+		}
+	}
+
+	/**
+	 * Calling this method will attempt delete the underlying snapshot directory recursively, if the state was
+	 * "ongoing". In this case, the state will be set to "deleted" as a result of this call.
+	 *
+	 * @return <code>true</code> if delete is successful, <code>false</code> otherwise.
+	 * @throws IOException if an exception happens during the delete.
+	 */
+	public boolean deleteIfIncompleteSnapshot() throws IOException {
+		return state.compareAndSet(State.ONGOING, State.DELETED) && fileSystem.delete(directory, true);
+	}
+
+	public boolean isSnapshotOngoing() {
+		return State.ONGOING.equals(state.get());
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()) {
+			return false;
+		}
+
+		SnapshotDirectory that = (SnapshotDirectory) o;
+
+		return directory.equals(that.directory);
+	}
+
+	@Override
+	public int hashCode() {
+		return directory.hashCode();
+	}
+
+	@Override
+	public String toString() {
+		return "SnapshotDirectory{" +
+			"directory=" + directory +
+			", state=" + state +
+			'}';
+	}
+
+	/**
+	 * Helper method that creates a temporary snapshot directory for the given path. This subclass of snapshot directory
+	 * will always return "null" as result of {@link #completeSnapshotAndGetHandle()} and can never reach the lifecycle
+	 * state "completed".
+	 */
+	public static SnapshotDirectory temporary(Path path) throws IOException {
+
+		return new SnapshotDirectory(path) {
```

Comment:
Not sure whether an anonymous class is strictly necessary here.

Reply:
I have changed that in my comments branch already. Made `SnapshotDirectory`  itself abstract in method `completeSnapshotAndGetHandle`, with two implementations (temporary, permanent) that have their static factory methods each.